### DISCUSSION TITLE
[Enhancement] [table-column] 表格列宽自适应

### DISCUSSION THREAD

**Discussion by iZaiZaiA** [element-plus/element-plus#18181] [Feature Request]
Feature type: **`Enhancement`**

## Details

- Enhanced component name: `table-column`

## Description

希望增加一个，列宽根据内容自适应宽度。

比如某一列的内容很多，很长，但具体有多长，无法得知，毕竟是动态的，但我又不想让其换行，而是横向显示，哪怕是出现横向滚动条也无所谓。

我不知道大佬们，能不能明白我的意思。

### Additional comments

![QQ20240905-094501](https://github.com/user-attachments/assets/61552524-caf1-4ff7-8ebc-669d07ce6f24)

**Comment by iZaiZaiA** [comment]
求大佬实现这个需求

**Comment by DDDDD12138** [comment]
可以尝试自己用 slot 去实现不换行，然后超出后出现滚动条：[Demo](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtdGFibGUgOmRhdGE9XCJ0YWJsZURhdGFcIiBzdHlsZT1cIndpZHRoOiAxMDAlXCI+XG4gICAgPGVsLXRhYmxlLWNvbHVtbiBwcm9wPVwiZGF0ZVwiIGxhYmVsPVwiRGF0ZVwiIHdpZHRoPVwiMTgwXCIgLz5cbiAgICA8ZWwtdGFibGUtY29sdW1uIHByb3A9XCJuYW1lXCIgbGFiZWw9XCJOYW1lXCIgd2lkdGg9XCIxODBcIiAvPlxuICAgIDxlbC10YWJsZS1jb2x1bW4gcHJvcD1cInRhZ1wiIGxhYmVsPVwiVGFnXCI+XG4gICAgICA8dGVtcGxhdGUgI2RlZmF1bHQ9XCJ7IHJvdzogeyB0YWdzID1bXSB9IH1cIj5cbiAgICAgICAgPGRpdiBjbGFzcz1cInRhZ3Mtd3JhcFwiPlxuICAgICAgICAgIDxlbC10YWcgdHlwZT1cImluZm9cIiB2LWZvcj1cInRhZyBpbiB0YWdzXCIgOmtleT1cInRhZ1wiPnt7IHRhZyB9fTwvZWwtdGFnPlxuICAgICAgICA8L2Rpdj5cbiAgICAgIDwvdGVtcGxhdGU+XG4gICAgPC9lbC10YWJsZS1jb2x1bW4+XG4gICAgPGVsLXRhYmxlLWNvbHVtbiBwcm9wPVwiYWN0aW9uXCIgbGFiZWw9XCJBY3Rpb25cIiB3aWR0aD1cIjE4MFwiIC8+XG4gIDwvZWwtdGFibGU+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IGxhbmc9XCJ0c1wiIHNldHVwPlxuY29uc3QgdGFibGVEYXRhID0gW1xuICB7XG4gICAgZGF0ZTogJzIwMTYtMDUtMDMnLFxuICAgIG5hbWU6ICdUb20nLFxuICAgIHRhZ3M6IFsnVGFnMScsICdUYWcyJywgJ1RhZzMnLCAnVGFnNCcsICdUYWc1JywgJ1RhZzYnXSxcbiAgICBhY3Rpb246ICdEZXRhaWwnLFxuICB9LFxuICB7XG4gICAgZGF0ZTogJzIwMTYtMDUtMDInLFxuICAgIG5hbWU6ICdUb20nLFxuICAgIHRhZ3M6IFsnVGFnMScsICdUYWcyJywgJ1RhZzMnLCAnVGFnNCcsICdUYWc1JywgJ1RhZzYnLCAnVGFnNycsICdUYWc4JywgJ1RhZzknXSxcbiAgICBhY3Rpb246ICdEZXRhaWwnLFxuICB9LFxuICB7XG4gICAgZGF0ZTogJzIwMTYtMDUtMDQnLFxuICAgIG5hbWU6ICdUb20nLFxuICAgIHRhZ3M6IFsnVGFnMScsICdUYWcyJ10sXG4gICAgYWN0aW9uOiAnRGV0YWlsJyxcbiAgfSxcbiAge1xuICAgIGRhdGU6ICcyMDE2LTA1LTAxJyxcbiAgICBuYW1lOiAnVG9tJyxcbiAgICB0YWdzOiBbJ1RhZzEnXSxcbiAgICBhY3Rpb246ICdEZXRhaWwnLFxuICB9LFxuXVxuPC9zY3JpcHQ+XG5cbjxzdHlsZT5cbi50YWdzLXdyYXAge1xuICBkaXNwbGF5OiBmbGV4O1xuICB3aGl0ZS1zcGFjZTogbm9yd2FycDtcbiAgb3ZlcmZsb3c6IGF1dG87XG4gIGdhcDogNnB4O1xufVxuPC9zdHlsZT5cbiIsImVsZW1lbnQtcGx1cy5qcyI6ImltcG9ydCBFbGVtZW50UGx1cyBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBnZXRDdXJyZW50SW5zdGFuY2UgfSBmcm9tICd2dWUnXG5cbmxldCBpbnN0YWxsZWQgPSBmYWxzZVxuYXdhaXQgbG9hZFN0eWxlKClcblxuZXhwb3J0IGZ1bmN0aW9uIHNldHVwRWxlbWVudFBsdXMoKSB7XG4gIGlmIChpbnN0YWxsZWQpIHJldHVyblxuICBjb25zdCBpbnN0YW5jZSA9IGdldEN1cnJlbnRJbnN0YW5jZSgpXG4gIGluc3RhbmNlLmFwcENvbnRleHQuYXBwLnVzZShFbGVtZW50UGx1cylcbiAgaW5zdGFsbGVkID0gdHJ1ZVxufVxuXG5leHBvcnQgZnVuY3Rpb24gbG9hZFN0eWxlKCkge1xuICBjb25zdCBzdHlsZXMgPSBbJ2h0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vZWxlbWVudC1wbHVzQGxhdGVzdC9kaXN0L2luZGV4LmNzcycsICdodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL2VsZW1lbnQtcGx1c0BsYXRlc3QvdGhlbWUtY2hhbGsvZGFyay9jc3MtdmFycy5jc3MnXS5tYXAoKHN0eWxlKSA9PiB7XG4gICAgcmV0dXJuIG5ldyBQcm9taXNlKChyZXNvbHZlLCByZWplY3QpID0+IHtcbiAgICAgIGNvbnN0IGxpbmsgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCdsaW5rJylcbiAgICAgIGxpbmsucmVsID0gJ3N0eWxlc2hlZXQnXG4gICAgICBsaW5rLmhyZWYgPSBzdHlsZVxuICAgICAgbGluay5hZGRFdmVudExpc3RlbmVyKCdsb2FkJywgcmVzb2x2ZSlcbiAgICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignZXJyb3InLCByZWplY3QpXG4gICAgICBkb2N1bWVudC5ib2R5LmFwcGVuZChsaW5rKVxuICAgIH0pXG4gIH0pXG4gIHJldHVybiBQcm9taXNlLmFsbFNldHRsZWQoc3R5bGVzKVxufVxuIiwidHNjb25maWcuanNvbiI6IntcbiAgXCJjb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IFwiRVNOZXh0XCIsXG4gICAgXCJqc3hcIjogXCJwcmVzZXJ2ZVwiLFxuICAgIFwibW9kdWxlXCI6IFwiRVNOZXh0XCIsXG4gICAgXCJtb2R1bGVSZXNvbHV0aW9uXCI6IFwiQnVuZGxlclwiLFxuICAgIFwidHlwZXNcIjogW1wiZWxlbWVudC1wbHVzL2dsb2JhbC5kLnRzXCJdLFxuICAgIFwiYWxsb3dJbXBvcnRpbmdUc0V4dGVuc2lvbnNcIjogdHJ1ZSxcbiAgICBcImFsbG93SnNcIjogdHJ1ZSxcbiAgICBcImNoZWNrSnNcIjogdHJ1ZVxuICB9LFxuICBcInZ1ZUNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogMy4zXG4gIH1cbn1cbiIsIlBsYXlncm91bmRNYWluLnZ1ZSI6IjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQgQXBwIGZyb20gJy4vQXBwLnZ1ZSdcbmltcG9ydCB7IHNldHVwRWxlbWVudFBsdXMgfSBmcm9tICcuL2VsZW1lbnQtcGx1cy5qcydcbnNldHVwRWxlbWVudFBsdXMoKVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPEFwcCAvPlxuPC90ZW1wbGF0ZT5cbiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcInZ1ZVwiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vQHZ1ZS9ydW50aW1lLWRvbUBsYXRlc3QvZGlzdC9ydW50aW1lLWRvbS5lc20tYnJvd3Nlci5qc1wiLFxuICAgIFwiQHZ1ZS9zaGFyZWRcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvc2hhcmVkQGxhdGVzdC9kaXN0L3NoYXJlZC5lc20tYnVuZGxlci5qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9lbGVtZW50LXBsdXNAbGF0ZXN0L2Rpc3QvaW5kZXguZnVsbC5taW4ubWpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXMvXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9lbGVtZW50LXBsdXNAbGF0ZXN0L1wiLFxuICAgIFwiQGVsZW1lbnQtcGx1cy9pY29ucy12dWVcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlQDIvZGlzdC9pbmRleC5taW4uanNcIlxuICB9LFxuICBcInNjb3Blc1wiOiB7fVxufSIsIl9vIjp7fX0=)

**Reply by iZaiZaiA** [reply]
首先，谢谢大佬，百忙之中抽出时间写一份demo来回答我的需求。

你尝试把左边的拉窄一点，会发现，列宽还是被限制了，我想要的是，在这种整体宽度不宽的情况下，列宽还是和内容最长的一致，后面的列，可以溢出，让表格出现整体的横向滚动条，而不是在列里出现滚动条，如果仅仅只是列里出现滚动条，我也不会提出这个需求了。

不知道你是否理解我的意思？

![QQ20241107-155930](https://github.com/user-attachments/assets/058605bb-f9fa-4fa8-b920-c56c00144e15)

![QQ20241107-160001](https://github.com/user-attachments/assets/2d865a3b-3404-4434-8836-2bae2ff91554)

**Reply by iZaiZaiA** [reply]
比如，tag列的最大宽度是400，我想实现，我不设置列宽400，就自动计算出宽度是400，并不论在什么情况下，都是400宽度。哪怕屏幕实际上很宽。因为这里只是比喻宽度400，因为动态数据问题，我不保证，每条数据的最大宽度400，也可能100，也可能900。意味着，我不能设置固定值，需要表格列自动的根据内容自适应，永远保证不换行的情况下，完整显示列的内容。

**Reply by DDDDD12138** [reply]
手动算宽度吧：[Demo](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtdGFibGUgOmRhdGE9XCJ0YWJsZURhdGFcIiBzdHlsZT1cIndpZHRoOiAxMDAlXCI+XG4gICAgPGVsLXRhYmxlLWNvbHVtbiBwcm9wPVwiZGF0ZVwiIGxhYmVsPVwiRGF0ZVwiIHdpZHRoPVwiMTgwXCIgLz5cbiAgICA8ZWwtdGFibGUtY29sdW1uIHByb3A9XCJuYW1lXCIgbGFiZWw9XCJOYW1lXCIgd2lkdGg9XCIxODBcIiAvPlxuICAgIDxlbC10YWJsZS1jb2x1bW4gcHJvcD1cInRhZ1wiIGxhYmVsPVwiVGFnXCIgOndpZHRoPVwidGFnc1dpZHRoXCI+XG4gICAgICA8dGVtcGxhdGUgI2RlZmF1bHQ9XCJ7IHJvdzogeyB0YWdzID1bXSB9LCAkaW5kZXggfVwiPlxuICAgICAgICA8ZGl2IGNsYXNzPVwidGFncy13cmFwXCIgOnJlZj1cImVsID0+IHRhZ3NDb250YWluZXJSZWZzWyRpbmRleF0gPSBlbFwiPlxuICAgICAgICAgIDxlbC10YWcgdHlwZT1cImluZm9cIiB2LWZvcj1cInRhZyBpbiB0YWdzXCIgOmtleT1cInRhZ1wiPnt7IHRhZyB9fTwvZWwtdGFnPlxuICAgICAgICA8L2Rpdj5cbiAgICAgIDwvdGVtcGxhdGU+XG4gICAgPC9lbC10YWJsZS1jb2x1bW4+XG4gICAgPGVsLXRhYmxlLWNvbHVtbiBwcm9wPVwiYWN0aW9uXCIgbGFiZWw9XCJBY3Rpb25cIiB3aWR0aD1cIjE4MFwiIC8+XG4gIDwvZWwtdGFibGU+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IGxhbmc9XCJ0c1wiIHNldHVwPlxuaW1wb3J0IHsgbmV4dFRpY2sgfSBmcm9tICd2dWUnO1xuaW1wb3J0IHsgcmVmIH0gZnJvbSAndnVlJ1xuXG5jb25zdCB0YWdzQ29udGFpbmVyUmVmcyA9IHJlZihbXSlcblxuY29uc3QgdGFnc1dpZHRoID0gcmVmKG51bGwpXG5cbmNvbnN0IHRhYmxlRGF0YSA9IFtcbiB7XG4gICAgZGF0ZTogJzIwMTYtMDUtMDMnLFxuICAgIG5hbWU6ICdUb20nLFxuICAgIHRhZ3M6IFsnVGFnMScsICdUYWcyJywgJ1RhZzMnLCAnVGFnNCcsICdUYWc1JywgJ1RhZzYnXSxcbiAgICBhY3Rpb246ICdEZXRhaWwnLFxuICB9LFxuICB7XG4gICAgZGF0ZTogJzIwMTYtMDUtMDInLFxuICAgIG5hbWU6ICdUb20nLFxuICAgIHRhZ3M6IFsnVGFnMScsICdUYWcyJywgJ1RhZzMnLCAnVGFnNCcsICdUYWc1JywgJ1RhZzYnLCAnVGFnNycsICdUYWc4JywgJ1RhZzknXSxcbiAgICBhY3Rpb246ICdEZXRhaWwnLFxuICB9LFxuICB7XG4gICAgZGF0ZTogJzIwMTYtMDUtMDQnLFxuICAgIG5hbWU6ICdUb20nLFxuICAgIHRhZ3M6IFsnVGFnMScsICdUYWcyJ10sXG4gICAgYWN0aW9uOiAnRGV0YWlsJyxcbiAgfSxcbiAge1xuICAgIGRhdGU6ICcyMDE2LTA1LTAxJyxcbiAgICBuYW1lOiAnVG9tJyxcbiAgICB0YWdzOiBbJ1RhZzEnXSxcbiAgICBhY3Rpb246ICdEZXRhaWwnLFxuICB9XG5dXG5cblxuXG5jb25zdCBzZXRUYWdzV2lkdGggPSBhc3luYyAoKSA9PiB7XG4gIGF3YWl0IG5leHRUaWNrKClcbiAgYXdhaXQgbmV4dFRpY2soKVxuICBhd2FpdCBuZXh0VGljaygpXG5cbiAgY29uc3QgcGFkZGluZ0FkZEdhcFdpZHRoID0gMjRcbiAgdGFnc1dpZHRoLnZhbHVlID0gTWF0aC5tYXgoLi4udGFnc0NvbnRhaW5lclJlZnMudmFsdWUubWFwKGVsID0+IGVsLnNjcm9sbFdpZHRoKSkgKyBwYWRkaW5nQWRkR2FwV2lkdGg7XG59XG5cbnNldFRhZ3NXaWR0aCgpXG4vLyBUTy1ETzog5b2T5pWw5o2uIHRhYmxlRGF0YSDmlLnlj5jml7YsIOmHjeaWsOiwg+eUqOWHveaVsOiuoeeul+WuveW6pu+8jOaXoOaVsOaNruaXtua4heepuuWuveW6plxuXG48L3NjcmlwdD5cblxuPHN0eWxlPlxuLnRhZ3Mtd3JhcCB7XG4gIGRpc3BsYXk6IGZsZXg7XG4gIHdoaXRlLXNwYWNlOiBub3J3YXJwO1xuICBvdmVyZmxvdzogYXV0bztcbiAgZ2FwOiA2cHg7XG59XG48L3N0eWxlPlxuIiwiZWxlbWVudC1wbHVzLmpzIjoiaW1wb3J0IEVsZW1lbnRQbHVzIGZyb20gJ2VsZW1lbnQtcGx1cydcbmltcG9ydCB7IGdldEN1cnJlbnRJbnN0YW5jZSB9IGZyb20gJ3Z1ZSdcblxubGV0IGluc3RhbGxlZCA9IGZhbHNlXG5hd2FpdCBsb2FkU3R5bGUoKVxuXG5leHBvcnQgZnVuY3Rpb24gc2V0dXBFbGVtZW50UGx1cygpIHtcbiAgaWYgKGluc3RhbGxlZCkgcmV0dXJuXG4gIGNvbnN0IGluc3RhbmNlID0gZ2V0Q3VycmVudEluc3RhbmNlKClcbiAgaW5zdGFuY2UuYXBwQ29udGV4dC5hcHAudXNlKEVsZW1lbnRQbHVzKVxuICBpbnN0YWxsZWQgPSB0cnVlXG59XG5cbmV4cG9ydCBmdW5jdGlvbiBsb2FkU3R5bGUoKSB7XG4gIGNvbnN0IHN0eWxlcyA9IFsnaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9lbGVtZW50LXBsdXNAbGF0ZXN0L2Rpc3QvaW5kZXguY3NzJywgJ2h0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vZWxlbWVudC1wbHVzQGxhdGVzdC90aGVtZS1jaGFsay9kYXJrL2Nzcy12YXJzLmNzcyddLm1hcCgoc3R5bGUpID0+IHtcbiAgICByZXR1cm4gbmV3IFByb21pc2UoKHJlc29sdmUsIHJlamVjdCkgPT4ge1xuICAgICAgY29uc3QgbGluayA9IGRvY3VtZW50LmNyZWF0ZUVsZW1lbnQoJ2xpbmsnKVxuICAgICAgbGluay5yZWwgPSAnc3R5bGVzaGVldCdcbiAgICAgIGxpbmsuaHJlZiA9IHN0eWxlXG4gICAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2xvYWQnLCByZXNvbHZlKVxuICAgICAgbGluay5hZGRFdmVudExpc3RlbmVyKCdlcnJvcicsIHJlamVjdClcbiAgICAgIGRvY3VtZW50LmJvZHkuYXBwZW5kKGxpbmspXG4gICAgfSlcbiAgfSlcbiAgcmV0dXJuIFByb21pc2UuYWxsU2V0dGxlZChzdHlsZXMpXG59XG4iLCJ0c2NvbmZpZy5qc29uIjoie1xuICBcImNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogXCJFU05leHRcIixcbiAgICBcImpzeFwiOiBcInByZXNlcnZlXCIsXG4gICAgXCJtb2R1bGVcIjogXCJFU05leHRcIixcbiAgICBcIm1vZHVsZVJlc29sdXRpb25cIjogXCJCdW5kbGVyXCIsXG4gICAgXCJ0eXBlc1wiOiBbXCJlbGVtZW50LXBsdXMvZ2xvYmFsLmQudHNcIl0sXG4gICAgXCJhbGxvd0ltcG9ydGluZ1RzRXh0ZW5zaW9uc1wiOiB0cnVlLFxuICAgIFwiYWxsb3dKc1wiOiB0cnVlLFxuICAgIFwiY2hlY2tKc1wiOiB0cnVlXG4gIH0sXG4gIFwidnVlQ29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiAzLjNcbiAgfVxufVxuIiwiUGxheWdyb3VuZE1haW4udnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCBBcHAgZnJvbSAnLi9BcHAudnVlJ1xuaW1wb3J0IHsgc2V0dXBFbGVtZW50UGx1cyB9IGZyb20gJy4vZWxlbWVudC1wbHVzLmpzJ1xuc2V0dXBFbGVtZW50UGx1cygpXG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8QXBwIC8+XG48L3RlbXBsYXRlPlxuIiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge1xuICAgIFwidnVlXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9AdnVlL3J1bnRpbWUtZG9tQGxhdGVzdC9kaXN0L3J1bnRpbWUtZG9tLmVzbS1icm93c2VyLmpzXCIsXG4gICAgXCJAdnVlL3NoYXJlZFwiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vQHZ1ZS9zaGFyZWRAbGF0ZXN0L2Rpc3Qvc2hhcmVkLmVzbS1idW5kbGVyLmpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXNcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL2VsZW1lbnQtcGx1c0BsYXRlc3QvZGlzdC9pbmRleC5mdWxsLm1pbi5tanNcIixcbiAgICBcImVsZW1lbnQtcGx1cy9cIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL2VsZW1lbnQtcGx1c0BsYXRlc3QvXCIsXG4gICAgXCJAZWxlbWVudC1wbHVzL2ljb25zLXZ1ZVwiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vQGVsZW1lbnQtcGx1cy9pY29ucy12dWVAMi9kaXN0L2luZGV4Lm1pbi5qc1wiXG4gIH0sXG4gIFwic2NvcGVzXCI6IHt9XG59IiwiX28iOnt9fQ==)

**Reply by iZaiZaiA** [reply]
也行吧，我其实是想着，能否把这个内置到表格组件里，通过列组件的属性来设置，是否要启用这个计算。

不管怎么样，还是要感谢大佬，非常感谢。

**Comment by 60late** [comment]
elementplus竟然没有这个功能，让我感觉很震惊
